### DISCUSSION TITLE
Questions regarding osi3::SensorViewConfiguration

### DISCUSSION THREAD

**Discussion by sebastian-wolter-itk** [OpenSimulationInterface/open-simulation-interface#779] [Q&A] [answered]
Hello everyone,

Question 1.)
[SensorViewConfiguration](https://opensimulationinterface.github.io/open-simulation-interface/structosi3_1_1SensorViewConfiguration.html) says: 
_"In response the environment simulation will configure the input and provide a new message of this type, which describes the actual configuration that it is going to employ"_

So the simulator sends a [SensorViewConfiguration](https://opensimulationinterface.github.io/open-simulation-interface/structosi3_1_1SensorViewConfiguration.html) message to the sensor model and the sensor model sends the same message back to simulator. The purpose is that the simulator does not know anything about the sensor models and need to ask for some info, so the sensorview message can be created accordingly.

But which signals are supposed to be filled by the sender and which by the receiver?
simulator: provides version, sensor_id, simulation_start_time
sensor model: provides all other signals ?

Question 2.)
is the [SensorViewConfiguration](https://opensimulationinterface.github.io/open-simulation-interface/structosi3_1_1SensorViewConfiguration.html) message supposed to be send per sensor? or just 1 for everything?

Or is it supposed to work like [SensorView](https://opensimulationinterface.github.io/open-simulation-interface/structosi3_1_1SensorView.html):
_"When simulating multiple distinct sensors, each sensor can consume an individual copy of the SensorView interface. This allows an independent treatment of the sensors. Alternatively combined sensor models can also consume one combined SensorView, with either combined or separate SensorData outputs, depending on model architecture."_

Does this also apply for SensorViewConfiguration? So both ways are allowed? If so, i would propose to add a similar text to the documentation. 

Regards,
Sebastian Wolter

**Comment by pmai** [comment, marked as answer]
Answering the questions back to front:

## Question 2

There is supposed to be a one to one correspondence between SensorViewConfiguration and SensorView: For each SensorView that a model wants to receive, it provides a SensorViewConfiguration request, which the simulator then uses to configure the corresponding SensorView. It provides a response SensorViewConfiguration to indicate what it will send, which can also be embedded into the SensorView.

So if a model wants to receive 3 SensorViews it will provide 3 SensorViewConfigurations, if it just wants one, it will provide one.

## Question 1

Both model and simulator will fill all fields (that are relevant, i.e. if no RadarSensorView is wanted/needed, no RadarSensorViewConfiguration sub-message is filled, etc.), just from a different perspective:

- The model will provide a SensorViewConfiguration as a **request**, that informs the simulator of what it wants to receive in terms of the SensorView.
- The simulator will provide a SensorViewConfiguration as a **response** that informs the model of what the simulator will in the end provide in terms of SensorView, given the constraints of what it can support.

If the response by the simulator is not sufficiently capable in terms of what the model requires it will cause an error or abort the simulation, or whatever is specified for the packaging layer of the model/simulator.

A more detailed discussion of this one-round-trip negotiation process can be seen in the OSMP specification that shows how this is supposed to be handled for FMUs as the packaging layer.

Hope this helps somewhat.

**Reply by sebastian-wolter-itk** [reply]
thanks for the quick feedback.
